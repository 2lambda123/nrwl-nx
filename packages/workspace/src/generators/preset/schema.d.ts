// nx-ignore-next-line
import type { Linter } from '@nrwl/linter';

import { Preset } from '../utils/presets';
import { PackageManager } from '@nrwl/devkit';

export interface Schema {
  name: string;
  npmScope?: string;
  style?: string;
  cli: string;
  linter?: Linter;
  preset: Preset;
  standaloneConfig?: boolean;
  packageManager?: PackageManager;
}
