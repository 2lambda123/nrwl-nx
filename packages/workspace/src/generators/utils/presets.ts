export enum Preset {
  Apps = 'apps',
  Empty = 'empty', // same as apps, deprecated
  Core = 'core', // same as npm, deprecated
  NPM = 'npm',
  TS = 'ts',
  WebComponents = 'web-components',
  Angular = 'angular',
  React = 'react',
  ReactExperimental = 'react-experimental',
  ReactNative = 'react-native',
  Expo = 'expo',
  NextJs = 'next',
  Nest = 'nest',
  Express = 'express',
}
