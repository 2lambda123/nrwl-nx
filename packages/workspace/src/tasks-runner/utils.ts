import { Task } from './tasks-runner';
import { ProjectMap, DependencyGraph } from '../command-line/shared';
import { createTask } from '../command-line/run-tasks/run-command';
import { Target } from '@angular-devkit/architect';

const commonCommands = ['build', 'test', 'lint', 'e2e', 'deploy'];

export function getPreconditions(
  tasks: Task[],
  graph: DependencyGraph
): Task[] {
  const preconditions = new Set<Task>();

  for (let task of tasks) {
    buildPreconditions(task.target, graph, preconditions);
  }

  return Array.from(preconditions);
}

function buildPreconditions(
  target: Target,
  graph: DependencyGraph,
  preconditions: Set<Task>
) {
  const { projects, dependencies } = graph;

  for (let dep of dependencies[target.project]) {
    let next: Target = { ...target, project: dep.projectName };
    buildPreconditions(next, graph, preconditions);

    const project = projects[dep.projectName];
    const task = createTask({
      project,
      target: next.target,
      configuration: next.configuration,
      overrides: next.overrides
    });

    preconditions.add(task);
  }
}

export function getCommand(cliCommand: string, isYarn: boolean, task: Task) {
  const args = Object.entries(task.overrides || {})
    .map(([prop, value]) => `--${prop}=${value}`)
    .join(' ');

  if (commonCommands.includes(task.target.target)) {
    const config = task.target.configuration
      ? `--configuration ${task.target.configuration} `
      : '';
    return `${cliCommand}${isYarn ? '' : ' --'} ${task.target.target} ${
      task.target.project
    } ${config} ${args}`.trim();
  } else {
    const config = task.target.configuration
      ? `:${task.target.configuration} `
      : '';
    return `${cliCommand}${isYarn ? '' : ' --'}  run ${task.target.project}:${
      task.target.target
    }${config} ${args}`.trim();
  }
}

export function getOutputs(p: ProjectMap, task: Task) {
  const architect = p[task.target.project].architect[task.target.target];
  let opts = architect.options || {};
  if (
    architect.configurations &&
    architect.configurations[task.target.configuration]
  ) {
    opts = {
      ...opts,
      ...architect.configurations[task.target.configuration]
    };
  }
  let outputs = [];
  if (opts.outputPath) {
    outputs.push(opts.outputPath);
  }
  return outputs;
}
