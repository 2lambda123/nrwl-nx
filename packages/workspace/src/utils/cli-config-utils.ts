import type { Tree } from '@angular-devkit/schematics';

import { readJsonInTree } from './ast-utils';

import type { NxJsonConfiguration } from '@nrwl/devkit';

/**
 * @deprecated Nx no longer supports workspace.json
 */
export function getWorkspacePath(host: Tree) {
  const possibleFiles = ['/angular.json', '/workspace.json'];
  return possibleFiles.filter((path) => host.exists(path))[0];
}

export function getNpmScope(host: Tree): string {
  return readJsonInTree<NxJsonConfiguration>(host, 'nx.json').npmScope;
}

export function parseTarget(targetString: string) {
  const [project, target, config] = targetString.split(':');
  return {
    project,
    target,
    config,
  };
}

export function editTarget(targetString: string, callback) {
  const parsedTarget = parseTarget(targetString);
  return serializeTarget(callback(parsedTarget));
}

export function serializeTarget({ project, target, config }) {
  return [project, target, config].filter((part) => !!part).join(':');
}
