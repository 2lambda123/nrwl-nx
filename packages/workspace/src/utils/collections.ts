/**
 * This file is used by `nx list --recommended` to list the recommended schematic collections (plugins)
 */

export enum Collection {
  Angular = '@nrwl/angular',
  Cypress = '@nrwl/cypress',
  Express = '@nrwl/express',
  Jest = '@nrwl/jest',
  Nest = '@nrwl/nest',
  Next = '@nrwl/next',
  Node = '@nrwl/node',
  React = '@nrwl/react',
  Storybook = '@nrwl/storybook',
  Web = '@nrwl/web'
}

export interface CollectionInfo {
  description: string;
  for?: string[];
  dependencies?: Collection[];
}

export type CollectionMap = {
  [collection in Collection]: CollectionInfo;
};

export const recommendedCollectionMap: CollectionMap = {
  [Collection.Angular]: {
    description:
      'Provides code generation for Angular, NgRx and DataPersistence',
    for: ['@angular/core'],
    dependencies: [Collection.Cypress, Collection.Jest]
  },
  [Collection.Cypress]: {
    description: 'Provides tools to run Cypress tests',
    for: ['@angular/core', 'react', '@nrwl/web', 'webpack']
  },
  [Collection.Express]: {
    description: 'Provides code generation and tools for Express',
    for: ['express'],
    dependencies: [Collection.Node, Collection.Jest]
  },
  [Collection.Jest]: {
    description: 'Provides tools to run Jest tests'
  },
  [Collection.Nest]: {
    description: 'Provides code generation and tools for NestJS',
    for: ['@nestjs/core'],
    dependencies: [Collection.Node, Collection.Jest]
  },
  [Collection.Next]: {
    description: 'Provides code generation and tools for Next.js',
    for: ['next'],
    dependencies: [Collection.React]
  },
  [Collection.Node]: {
    description: 'Provides code generation and tools for Node'
  },
  [Collection.React]: {
    description: 'Provides code generation and tools for React',
    for: ['react'],
    dependencies: [Collection.Cypress, Collection.Jest, Collection.Web]
  },

  [Collection.Storybook]: {
    description: 'Provides tools to run Storybook tests',
    for: ['@storybook/core'],
    dependencies: [Collection.Cypress]
  },
  [Collection.Web]: {
    description:
      'Provides code generation and tools for standard Web applications'
  }
};
