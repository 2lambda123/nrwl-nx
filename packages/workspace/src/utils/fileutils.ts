import {
  createReadStream,
  createWriteStream,
  existsSync,
  writeFileSync,
  mkdirSync,
  renameSync as fsRenameSync,
  statSync,
} from 'fs';
import { basename, dirname, resolve } from 'path';
import { serializeJson, readJsonFile, writeJsonFile } from '@nrwl/devkit';

export { serializeJson, readJsonFile, writeJsonFile };

export function writeToFile(filePath: string, str: string) {
  mkdirSync(dirname(filePath), { recursive: true });
  writeFileSync(filePath, str);
}

/**
 * This method is specifically for updating a JSON file using the filesystem
 *
 * @remarks
 * If you are looking to update a JSON file in a tree, look for ./ast-utils#updateJsonInTree
 * @param path Path of the JSON file on the filesystem
 * @param callback Manipulation of the JSON data
 */
export function updateJsonFile(path: string, callback: (a: any) => any) {
  const json = readJsonFile(path);
  callback(json);
  writeJsonFile(path, json);
}

export function copyFile(file: string, target: string) {
  const f = basename(file);
  const source = createReadStream(file);
  const dest = createWriteStream(resolve(target, f));
  source.pipe(dest);
  source.on('error', (e) => console.error(e));
}

export function directoryExists(name) {
  try {
    return statSync(name).isDirectory();
  } catch (e) {
    return false;
  }
}

export function fileExists(filePath: string): boolean {
  try {
    return statSync(filePath).isFile();
  } catch (err) {
    return false;
  }
}

export function renameSync(
  from: string,
  to: string,
  cb: (err: Error | null) => void
) {
  try {
    if (!existsSync(from)) {
      throw new Error(`Path: ${from} does not exist`);
    } else if (existsSync(to)) {
      throw new Error(`Path: ${to} already exists`);
    }

    // Make sure parent path exists
    const parentPath = resolve(to, '..');
    mkdirSync(parentPath, { recursive: true });

    fsRenameSync(from, to);
    cb(null);
  } catch (e) {
    cb(e);
  }
}

export function isRelativePath(path: string): boolean {
  return (
    path === '.' ||
    path === '..' ||
    path.startsWith('./') ||
    path.startsWith('../')
  );
}
