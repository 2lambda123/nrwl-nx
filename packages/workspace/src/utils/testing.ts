import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import type { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/express',
  join(__dirname, '../../../express/generators.json')
);

testRunner.registerCollection(
  '@nrwl/react',
  join(__dirname, '../../../react/generators.json')
);

testRunner.registerCollection(
  '@nrwl/angular',
  join(__dirname, '../../../angular/generators.json')
);

testRunner.registerCollection(
  '@nrwl/next',
  join(__dirname, '../../../next/generators.json')
);

testRunner.registerCollection(
  '@nrwl/node',
  join(__dirname, '../../../node/generators.json')
);

testRunner.registerCollection(
  '@nrwl/nest',
  join(__dirname, '../../../nest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/web',
  join(__dirname, '../../../web/generators.json')
);

/**
 * @deprecated This will be removed in v17. Prefer writing Nx Generators with @nrwl/devkit. Generators can be tested by simply calling them in it().
 */
export function runSchematic<T extends object = any>(
  schematicName: string,
  options: T,
  tree: Tree
) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

/**
 * @deprecated This will be removed in v17. Prefer writing Nx Generators with @nrwl/devkit. Generators can be tested by simply calling them in it().
 */
export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}
