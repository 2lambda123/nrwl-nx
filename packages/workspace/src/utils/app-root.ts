import { fileExists } from './fileutils';
import * as path from 'path';

function normalizeBackslashes(dir) {
	return dir.replace(/\\/g,'/')
}

export const appRootPath = normalizeBackslashes(pathInner(__dirname));

function pathInner(dir: string): string {
  if (path.dirname(dir) === dir) return process.cwd();
  if (
    fileExists(path.join(dir, 'workspace.json')) ||
    fileExists(path.join(dir, 'angular.json'))
  ) {
    return dir;
  } else {
    return pathInner(path.dirname(dir));
  }
}
