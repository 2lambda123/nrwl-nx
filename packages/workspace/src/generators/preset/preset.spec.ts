import { readProjectConfiguration, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { overrideCollectionResolutionForTesting } from '@nx/devkit/ngcli-adapter';
import { presetGenerator } from './preset';
import * as path from 'path';
import { Preset } from '../utils/presets';

describe('preset', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    overrideCollectionResolutionForTesting({
      '@nx/workspace': path.join(
        __dirname,
        '../../../../workspace/generators.json'
      ),
      '@nx/angular': path.join(
        __dirname,
        '../../../../angular/generators.json'
      ),
      '@nx/linter': path.join(__dirname, '../../../../linter/generators.json'),
      '@nx/nest': path.join(__dirname, '../../../../nest/generators.json'),
      '@nx/node': path.join(__dirname, '../../../../node/generators.json'),
      '@nx/jest': path.join(__dirname, '../../../../jest/generators.json'),
      '@nx/cypress': path.join(
        __dirname,
        '../../../../cypress/generators.json'
      ),
      '@nx/express': path.join(
        __dirname,
        '../../../../express/generators.json'
      ),
    });
  });

  afterEach(() => {
    overrideCollectionResolutionForTesting(null);
  });

  it(`should create files (preset = ${Preset.AngularMonorepo})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.AngularMonorepo,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.children('apps/proj')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/app')).toMatchSnapshot();
  }, 20000);

  it(`should create files (preset = ${Preset.WebComponents})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.WebComponents,
    });
    expect(tree.exists('/apps/proj/src/main.ts')).toBe(true);
  });

  it(`should create files (preset = ${Preset.ReactMonorepo})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactMonorepo,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('/apps/proj/src/main.tsx')).toBe(true);
    expect(readProjectConfiguration(tree, 'proj').targets.serve).toBeDefined();
  });

  it(`should create files (preset = ${Preset.NextJs})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.NextJs,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('/apps/proj/pages/index.tsx')).toBe(true);
  });

  it(`should create files (preset = ${Preset.Express})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.Express,
      linter: 'eslint',
    });

    expect(tree.exists('apps/proj/src/main.ts')).toBe(true);
    expect(tree.exists('apps/proj/.eslintrc.json')).toBe(true);
  });

  it('should create files (preset = react-native)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactNative,
      linter: 'eslint',
    });

    expect(tree.exists('/apps/proj/src/app/App.tsx')).toBe(true);
  });

  it(`should create files (preset = ${Preset.ReactStandalone} bundler = webpack)`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactStandalone,
      style: 'css',
      linter: 'eslint',
      bundler: 'webpack',
    });
    expect(tree.exists('webpack.config.js')).toBe(true);
    expect(
      readProjectConfiguration(tree, 'proj').targets.serve
    ).toMatchSnapshot();
  });

  it(`should create files (preset = ${Preset.ReactStandalone} bundler = vite)`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactStandalone,
      style: 'css',
      linter: 'eslint',
      bundler: 'vite',
    });
    expect(tree.exists('vite.config.ts')).toBe(true);
    expect(
      readProjectConfiguration(tree, 'proj').targets.serve
    ).toMatchSnapshot();
  });
});
