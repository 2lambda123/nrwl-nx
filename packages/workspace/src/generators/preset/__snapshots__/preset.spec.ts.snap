// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset should create files (preset = angular-monorepo) 1`] = `
Array [
  "tsconfig.app.json",
  "tsconfig.spec.json",
  "src",
  "tsconfig.editor.json",
  "tsconfig.json",
  ".eslintrc.json",
  "jest.config.ts",
]
`;

exports[`preset should create files (preset = angular-monorepo) 2`] = `
Array [
  "favicon.ico",
  "index.html",
  "main.ts",
  "styles.css",
  "assets",
  "app",
  "test-setup.ts",
]
`;

exports[`preset should create files (preset = angular-monorepo) 3`] = `
Array [
  "app.module.ts",
  "app.component.css",
  "app.component.html",
  "app.component.spec.ts",
  "app.component.ts",
  "nx-welcome.component.ts",
]
`;

exports[`preset should create files (preset = react-standalone bundler = vite) 1`] = `
Object {
  "configurations": Object {
    "development": Object {
      "buildTarget": "proj:build:development",
      "hmr": true,
    },
    "production": Object {
      "buildTarget": "proj:build:production",
      "hmr": false,
    },
  },
  "defaultConfiguration": "development",
  "executor": "@nrwl/vite:dev-server",
  "options": Object {
    "buildTarget": "proj:build",
  },
}
`;

exports[`preset should create files (preset = react-standalone bundler = webpack) 1`] = `
Object {
  "configurations": Object {
    "development": Object {
      "buildTarget": "proj:build:development",
    },
    "production": Object {
      "buildTarget": "proj:build:production",
      "hmr": false,
    },
  },
  "defaultConfiguration": "development",
  "executor": "@nrwl/webpack:dev-server",
  "options": Object {
    "buildTarget": "proj:build",
    "hmr": true,
  },
}
`;
